src/main/scala/me/rexim/morganey/ReplAutocompletion.scala

```diff
@@ -0,0 +1,153 @@
+package me.rexim.morganey
+
+import java.io.File
+import java.util.{List => Jlist}
+
+import jline.console.completer.Completer
+import me.rexim.morganey.ast.MorganeyBinding
+import me.rexim.morganey.module.ModuleFinder
+
+import scala.annotation.tailrec
+import scala.util.Try
+
+class ReplAutocompletion(globalContext: () => List[MorganeyBinding]) extends Completer {
+
+  override def complete(buffer: String, cursor: Int, candidates: Jlist[CharSequence]): Int = {
+    val knownVariableNames = globalContext().map(_.variable.name)
+    val (beforeCursor, _) = buffer splitAt cursor
+    lazy val definitions = matchingDefinitions(beforeCursor, knownVariableNames)
+
+    def autoCompleteWith(xs: List[String]) = {
+      for (elem <- xs) candidates.add(elem)
+      if (candidates.isEmpty) -1 else 0
+    }
+
+    beforeCursor match {
+      // if a load statement was typed in (potential partially)
+      case LoadStatement(parts, endsWithDot) =>
+        val autocompletedModules = autocompleteLoadStatement(parts, endsWithDot)
+        autoCompleteWith(autocompletedModules map (m => s"load $m"))
+      // if there are no names, REPL can't autocomplete
+      case _ if knownVariableNames.isEmpty   => -1
+      // if there are definitions matching the users input, use them for completion
+      case _ if definitions.nonEmpty         => autoCompleteWith(definitions)
+      // if nothing was typed into the repl, autocomplete with all known names
+      case _ if buffer.trim.isEmpty          => autoCompleteWith(knownVariableNames)
+      // if something was typed into the repl, autocomplete with all names starting with the input text
+      case _                                 =>
+        autoCompleteWith(knownVariableNames filter (_.toLowerCase startsWith beforeCursor.toLowerCase))
+    }
+  }
+
+  private def autocompleteLoadStatement(parts: List[String], endsWithDot: Boolean): List[String] = {
+    import MorganeyInterpreter.moduleFinder
```

Comment:
Could you please provide the module finder via constructor of the ReplAutocompletion class as you did for the global context? It's actually my fault. I should've noticed that the load expression modifies the REPL state and therefore should be a part of the global context.

I'll create a separate refactoring issue for that. We will need to define the global context entity as the list of binding plus the module finder, because as we add more feature the notion of the global context will extend further.